cmake: version tweak from git

JordanYates opened the pull request:
Add the option for the version tweak field to be populated automatically from the git commit hash.

I find the default `major.minor.revision` is sufficient granularity for user provided versioning.
This option lets the tweak contain useful information automatically (the git commit hash of the build).

tejlmand reviewed (changes requested):
The main purpose of version is to ensure that versioning of Zephyr itself works, but having this code generic so that it's reusable for applications is a plus.

One of the usage is looking up correct Zephyr as part of `find_package(Zephyr [version])`, thus we shouldn't require any external tools as part of this.

As Zephyr's version infrastructure is there and is useful for applications, then a small enhancement to support git sha in tweak field instead of a fixed number seems reasonable to me, but I would prefer that such customization is not done inside the VERSION file itself (as proposed in this PR with the `git` value).

Perhaps allowing the build system to overrule / specify how APP_TWEAK is obtained through CMake (like it can be done for `BUILD_VERSION`) or with a Kconfig is a better approach.

And / or in combination with that, adjust the `gen_version_h.cmake` to use the provided information to decide if the tweak part should be constructed using `git`, or fetched directly from VERSION file. (Perhaps code can be adjusted to allow a missing / empty tweak field)

JordanYates commented:
Updated implementation based on @tejlmand feedback. It is now enabled through a kconfig option, and all the logic except the value override has been moved outside of `version.cmake`.

marc-hb commented on CMakeLists.txt:
```diff
@@ -559,12 +559,36 @@ add_custom_command(
+    find_package(Git QUIET)
+    if(GIT_FOUND)
+      execute_process(
+        COMMAND ${GIT_EXECUTABLE} rev-parse HEAD
```
Can you move this code to `cmake/gen_version_h.cmake`, where `git describe` already is? So all git commands are in the same place, otherwise it's confusing and very time-consuming to find which git command does what, when and where.

Also, it's not like the top-level `CMakeList.txt` file is too short...

JordanYates replied:
Good point, move the logic into `gen_version_h.cmake`.

marc-hb reviewed (commented):
There are so many version fields that they make my head spin every time: major, minor, patchlevel, tweak, string, number, "code", extended, kernel, app, hex, decimal, "extra",... The number of fields seems to grow a bit out of control to be honest. So apologies for any stupid question and thanks in advance for "re-interpreting" them correctly :-)


A long time ago, I tried to summarize stuff in commit a880fb134326 at the top of version.cmake and there have been a couple updates since but I'm afraid this needs another, hopefully small doc refresh. For instance, the top of version.cmake refers to a `git.cmake` file that does not exist any more. Can you please also add a reference to docs/build/version/index.rst there? I this didn't exist at the time, so maybe there is some duplication between the two now. BTW it's really weird that this page lists `$(VERSION_MAJOR)` for Kconfig, not `$(APP_VERSION_MAJOR)`.

I'm sorry to do the typical thing and ask for a "clean-up" when all you want is to add your small "tweak" but I'm sure it will really help review and expedite your change, so it is also in your own interest? :-) Also, I'm not asking for any code change right now; just some clarifications and fixups in comments and docs.

marc-hb commented on cmake/modules/version.cmake:
```diff
@@ -66,6 +66,10 @@ foreach(type file IN ZIP_LISTS VERSION_TYPE VERSION_FILE)
 
+  if(VERSION_TWEAK_OVERRIDE)
+    set(${type}_VERSION_TWEAK ${VERSION_TWEAK_OVERRIDE})
+  endif()
```
Wait: why not add the same feature for the kernel too? Even if it ends up rarely ever used, simplicity and symmetry should globally results in LESS code easier to read.

JordanYates replied:
It will run for any user of `version.cmake` that sets the define, that APP is the only one to currently do so is not relevant to this code.

The option is of less use for the kernel IMO, whereas the application version is automatically used to sign MCUboot applications

JordanYates commented on tests/lib/app_version/git_tweak/VERSION:
```diff
@@ -0,0 +1,5 @@
+VERSION_MAJOR = 2
+VERSION_MINOR = 3
+PATCHLEVEL = 4
+VERSION_TWEAK = 0
```
This test is an odd one because we don't have a way to determine what value VERSION_TWEAK should end up with inside the test itself, as it will change with the git commit. Instead need to test the values it is not.

I don't care whether it is 0 because it was explicitly set to 0 here, or whether it was set to 0 because something was undefined, I only care that the final value is NOT 0.

JordanYates commented on doc/build/version/index.rst:
```diff
@@ -188,6 +188,30 @@ The following variable are available for usage in CMake files:
+   # Generated version file
+   > cat app_version.h
+   ...
+   #define APP_TWEAK                    1683891135
```
The CMake variable is `APP_VERSION_TWEAK`, the field in `app_version.h` is `VERSION_TWEAK`.
The reason it is in decimal because that is the only form accepted by the MCUboot signing tools, which consumes the value.

JordanYates commented on tests/lib/app_version/git_tweak/VERSION:
```diff
@@ -0,0 +1,5 @@
+VERSION_MINOR = 3
+PATCHLEVEL = 4
+VERSION_TWEAK = 0
+EXTRAVERSION = tweak
```
Changed the value and added the validation to the test

JordanYates commented on doc/build/version/index.rst:
```diff
@@ -188,6 +188,30 @@ The following variable are available for usage in CMake files:
+
+   # APP_TWEAK is equivalent to hash
+   > printf "%08x" 1683891135
+   645e23bf
```
This is a good point, `APPVERSION` is constructed assuming that `APP_TWEAK` can only be an 8 bit number, but there is no validation on this parameter and when the value is used to sign MCUboot binaries that is a 32 bit value.

The documentation of this value actually states that the LSB is reserved for future use, not set to `KERNEL_TWEAK`:
https://github.com/zephyrproject-rtos/zephyr/blob/66630961627f4d8625e2dbe1cb7afe4abd690b2e/include/zephyr/kernel_version.h#L24-L31

JordanYates commented:
As the new functionality won't make 3.7, I have made an alternate PR with just the doc cleanup and the basic test application which should be able to make it in, #7435.

JordanYates committed:
cmake: git.cmake: common git helpers

Move generic git functions into a dedicated file under `cmake/`.

Signed-off-by: Jordan Yates <jordan@embeint.com>

JordanYates committed:
cmake: git.cmake: helpers for working with hashes

Add two git helpers, the first for getting the git commit hash of the
latest commit in a directory, the second to shorten it to a form usable
in a `uint32_t`.

Signed-off-by: Jordan Yates <jordan@embeint.com>

JordanYates committed:
cmake: gen_version_h: add commit hash as integer

Include the first 32 bits of the git commit hash into the generated
version files in integer form. This can be an easier form to work with
than the `APP_BUILD_VERSION` define, which includes a superset of the
same information but in string form.

Signed-off-by: Jordan Yates <jordan@embeint.com>

JordanYates committed:
cmake: kconfig.cmake: inject application git commit

Inject application git commit information into Kconfig.

Signed-off-by: Jordan Yates <jordan@embeint.com>

JordanYates committed:
modules: mcuboot: automatic build from git

Add an option to automatically override the `build` version number with
the numeric equivalent of the first 8 characters of the git commit hash.

Signed-off-by: Jordan Yates <jordan@embeint.com>

JordanYates committed:
tsets: cmake: app_version: test `APP_GIT_COMMIT_HASH_SHORT`

Test the generation of `APP_GIT_COMMIT_HASH_SHORT` by comparing it to
the value of `APP_BUILD_VERSION`.

Signed-off-by: Jordan Yates <jordan@embeint.com>

JordanYates commented:
Alternate implementation of the same end behavior in #76133
